docs/building_pipelines.rst

```diff
@@ -293,3 +294,423 @@ Warnings raised will also be logged to the error log but will not cause data to
 
 See :ref:`AccessingLogs`: for details have how to access the system logs.
 
+Debugging & Tracing Pipelines
+-----------------------------
+
+Fledge has a feature that allows the debugging of pipelines in south and north services. It provides a mechanism to view the data as it flows through the pipeline.
+
+The debugger is designed to show the data as it traverses the pipeline within the service. Users may:
+
+   - Show the data at each stage in the pipeline.
+
+   - Configure the number of readings to store at each stage in the pipeline.
+
+   - Pause the ingest into the pipeline.
+
+   - Suspend the output of the pipeline.
+
+   - Ingest a number of readings into the paused pipeline.
+
+   - Replay the currently buffered readings for a pipeline.
+
+   - Resume flow into and out of the pipeline.
+
+
+The *replay* operation is useful when manipulating the configuration of the pipeline components. The user may change a filter configuration and replay the saved readings to see the impact of the configuration change. This may be repeated multiple times until the user is satisfied with the result of the filter configuration.
+
+Command Line Interface
+~~~~~~~~~~~~~~~~~~~~~~
+
+A command line interface is provided to access the pipeline debugger. The interface can be started by running the *fledge* script that is used to start, stop and monitor Fledge or by calling the *scripts/debug/debug* script directly.
+
+.. note::
+
+   Due to the sensitive nature of pipeline debugging only user with edit permissions may use this interface. Additionally authentication must be enabled on the Fledge instance for this interface to work.
+
+Starting the Debugger
+#####################
+
+The *debug* command should be given the name of a south or north service to debug.
+
+.. code-block:: bash
+
+   $ scripts/fledge -u admin debug Lathe
+   Debug: Lathe$ 
+
+Upon starting a new prompt will be displayed, if the service is a valid service to be debugged and is running. The user may now enter the various debug commands at this prompt. To complete the debugging session type *exit* or Control-D.
+
+To see a list of available commands enter the command *commands*
+
+.. code-block:: bash
+
+   Debug: Lathe$ commands
+   attach:              Attach the pipeline debugger
+   buffer:              Return the contents of the buffers at every pipeline element
+   detach:              Detach the debugger from the pipeline
+   isolate:             Isolate the pipeline from the destination
+   replay:              Replay the buffered data through the pipeline
+   resumeIngest:        Resume the flow of data into the pipeline
+   setBuffer:           Set the number of readings to hold in each buffer, passing an integer argument
+   state:               Return the state of the debugger
+   step:                Allow readings to flow into the pipeline. Passed an optional number of readings to ingest; default to 1 if omitted
```

Comment:
Is it possible for the number of steps to exceed the buffer size?
We should document the relationship between steps and buffer size?